### PR TITLE
User recommendation

### PR DESCRIPTION
**Pull request by itsStrobe**
Uses the K Means clustering algorithm, Stanford's Tagger and Austin Chen's logistic regression model to cluster users by similarities on the messages sent through the platform. This clustering is made in order to give the users a list of possible matches with users that share common conversations.

### CODE REVIEW

**Review by acolwell — COMMENTED** [review]
Please make this relative to rebased_db_integration. Otherwise it looks like the pull request includes everything changed in that branch AND your new changes. Doing that makes it difficult to review.

File: src/codeu/chat/codeU_db/DataBaseConnection.java [hunk 1]
```diff
@@ -4,6 +4,7 @@
 import codeu.chat.common.Message;
 import codeu.chat.common.User;
 import codeu.chat.util.*;
+import org.omg.CORBA.PUBLIC_MEMBER;
```
Comment (acolwell) [inline src/codeu/chat/codeU_db/DataBaseConnection.java]: what is this for? Remove?

File: src/codeu/chat/relay/ServerFrontEnd.java [hunk 2]
```diff
@@ -35,72 +35,72 @@
...
-      final Time time = Time.SERIALIZER.read(in);
-
-      // I could have passed the relay and use its "pack" method but that would
-      // have been more work than just building an object here.
-      return new Relay.Bundle.Component() {
-        @Override
-        public Uuid id() { return id; }
         @Override
-        public String text() { return text; }
+        public Relay.Bundle.Component read(InputStream in) throws IOException {
```
Comment (acolwell) [inline src/codeu/chat/relay/ServerFrontEnd.java:39]: It looks like the changes in this file are whitespace only. If that's correct, please revert.

File: src/codeu/chat/server/Model.java [hunk 3]
```diff
@@ -680,4 +713,67 @@ public Uuid conversationID(Uuid message) {
     String query = "SELECT CONVERSATIONID FROM MESSAGES WHERE ID = ?;";
     return dbConnection.getConversationID(parameters, query);
   }
+
+  public void add(Uuid user) {
```
Comment (acolwell) [inline src/codeu/chat/server/Model.java]: Rename this to addCluster() or something similar? Also the parameter name should be something like cluserId I think if I'm understanding the code correctly.

File: src/codeu/chat/server/Model.java [hunk 4]
```diff
@@ -680,4 +713,67 @@ public Uuid conversationID(Uuid message) {
...
+      LOG.info("User cluster added: " + user);
+    } catch (Exception e) {
+      LOG.info(
+          "newUserCluster fail - Verify connection and try again shortly");
+      System.err.println(e.getClass().getName() + ": " + e.getMessage());
+      System.exit(0);
+    }
+  }
+
+  public void update(Uuid user, int cluster) {
```
Comment (acolwell) [inline src/codeu/chat/server/Model.java]: Rename method to be a little more descriptive. Perhaps something like assignUserToCluster() ?

File: src/codeu/chat/server/RemoteRelay.java [hunk 5]
```diff
@@ -59,63 +59,63 @@ public Component(Uuid id, Time time, String text) {
   private static final Serializer<Relay.Bundle.Component> COMPONENT_SERIALIZER =
       new Serializer<Relay.Bundle.Component>() {
 
-    @Override
-    public Relay.Bundle.Component read(InputStream in) throws IOException {
+        @Override
```
Comment (acolwell) [inline src/codeu/chat/server/RemoteRelay.java:62]: Looks like all the changes in this file are whitespace only. If so, please revert.

File: src/codeu/chat/server/View.java [hunk 6]
```diff
@@ -33,6 +33,8 @@
 import codeu.chat.util.Uuid;
 import codeu.chat.util.store.StoreAccessor;
 
+import javax.jws.soap.SOAPBinding;
```
Comment (acolwell) [inline src/codeu/chat/server/View.java]: remove?

File: src/codeu/chat/util/Logger.java [hunk 7]
```diff
@@ -88,25 +88,25 @@ public void info(String message, Object... params) {
       @Override
       public void warning(String message, Object... params) {
         log.logp(java.util.logging.Level.WARNING,
-                 Thread.currentThread().getStackTrace()[2].getClassName(),
-                 Thread.currentThread().getStackTrace()[2].getMethodName(),
-                 String.format(message, params));
+            Thread.currentThread().getStackTrace()[2].getClassName(),
```
Comment (acolwell) [inline src/codeu/chat/util/Logger.java:91]: looks like whitespace only changes. Please revert.

File: src/codeu/chat/server/user_recommendation/K_Means.java [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+      for (int i = 0; i < numClusters; i++) {
+        Mood avg = Mood.average(sums[i], counter[i]);
+        clusterVector.get(i).updateEntry(keyWord, avg);
+      }
+    }
+  }
+
+  public boolean runClusterer(int iterations) {
+    System.out.println("Running Clusterer");
+    classifier.trainModel();
```
Comment (acolwell) [inline src/codeu/chat/server/user_recommendation/K_Means.java]: Do you really need to train the model everytime?

File: src/codeu/chat/server/user_recommendation/K_Means.java [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+    String[] tags = tagger.tag(tokens);
+
+    POSSample sample = new POSSample(tokens, tags);
+
+    taggedSentence = sample.toString();
+
+    return taggedSentence;
+  }
+
+  private void InitializeClusters() {
```
Comment (acolwell) [inline src/codeu/chat/server/user_recommendation/K_Means.java]: Make first char lower case here and elseqhere to be consistent with other function names.

File: test/codeu/chat/relay/ServerTest.java [hunk 10]
```diff
@@ -12,17 +12,17 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package codeu.chat.relay;
+    package codeu.chat.relay;
```
Comment (acolwell) [inline test/codeu/chat/relay/ServerTest.java]: Looks like all the changes in this file are whitespace changes. Please revert.

**Review by acolwell — APPROVED** [review]
LGTM

File: src/codeu/chat/relay/ServerFrontEnd.java (hunk 2 above)
Comment (itsStrobe) [reply src/codeu/chat/relay/ServerFrontEnd.java:39]: I don't know why this is signaling whitespace differences when it is the same file used in [Google's master branch](https://github.com/google/codeu_project_2017/blob/master/src/codeu/chat/relay/ServerFrontEnd.java). I suppose that due to IntelliJ's reformatting there was a change in a branch that is now showing the different whitespaces.

File: src/codeu/chat/server/RemoteRelay.java (hunk 5 above)
Comment (itsStrobe) [reply src/codeu/chat/server/RemoteRelay.java:62]: Same issue as with ServerFrontEnd.java.

File: src/codeu/chat/util/Logger.java (hunk 7 above)
Comment (itsStrobe) [reply src/codeu/chat/util/Logger.java:91]: Same issue as with ServerFrontEnd.java.

**Pull request merged**